File: ios/RCTTwilioChat/RCTTwilioChatChannels.m
//
//  RCTTCHChannels.m
//  TwilioIPExample
//
//  Created by Brad Bumbalough on 6/2/16.
//  Copyright © 2016 Facebook. All rights reserved.
//

#import "RCTTwilioChatChannels.h"
#import "RCTTwilioChatClient.h"
#import "RCTConvert+TwilioChatClient.h"
#import "RCTTwilioChatPaginator.h"
#import <React/RCTUtils.h>

@implementation RCTTwilioChatChannels

RCT_EXPORT_MODULE();

+ (void)loadChannelFromSid:(NSString *)sid :(void (^)(TCHResult *result, TCHChannel *channel))completion {
  TwilioChatClient *client = [[RCTTwilioChatClient sharedManager] client];
    [[client channelsList] channelWithSidOrUniqueName:sid completion:completion];
}

#pragma mark Channel Methods

RCT_REMAP_METHOD(getUserChannels, userChannels_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    TwilioChatClient *client = [[RCTTwilioChatClient sharedManager] client];
    [[client channelsList] userChannelDescriptorsWithCompletion:^(TCHResult *result, TCHChannelDescriptorPaginator *paginator) {
        if (result.isSuccessful) {
            NSString *uuid = [RCTTwilioChatPaginator setPaginator:paginator];
            resolve(@{
                      @"sid":uuid,
                      @"type": @"ChannelDescriptor",
                      @"paginator": [RCTConvert TCHChannelDescriptorPaginator:paginator]
                      });

        }
        else {
            reject(@"get-user-channels-error", @"Error occured while attempting to get the user channels.", result.error);
        }
    }];
}

RCT_REMAP_METHOD(getPublicChannels, publicChannels_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    TwilioChatClient *client = [[RCTTwilioChatClient sharedManager] client];
    [[client channelsList] publicChannelDescriptorsWithCompletion:^(TCHResult *result, TCHChannelDescriptorPaginator *paginator) {
        if (result.isSuccessful) {
            NSString *uuid = [RCTTwilioChatPaginator setPaginator:paginator];
            resolve(@{
                      @"sid":uuid,
                      @"type": @"ChannelDescriptor",
                      @"paginator": [RCTConvert TCHChannelDescriptorPaginator:paginator]
                      });

        }
        else {
            reject(@"get-public-channels-error", @"Error occured while attempting to get the public channels.", result.error);
        }
    }];
}

RCT_REMAP_METHOD(getSubscribedChannels, subscribedChannels_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  TwilioChatClient *client = [[RCTTwilioChatClient sharedManager] client];
  NSArray<TCHChannel *>* channels = [[client channelsList] subscribedChannels];
  resolve([RCTConvert TCHChannels:channels]);
}


RCT_REMAP_METHOD(createChannel, options:(NSDictionary *)options create_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  TwilioChatClient *client = [[RCTTwilioChatClient sharedManager] client];
  [[client channelsList] createChannelWithOptions:options completion:^(TCHResult *result, TCHChannel *channel) {
    if (result.isSuccessful) {
      resolve([RCTConvert TCHChannel:channel]);
    }
    else {
      reject(@"create-error", @"Error occured while creating a new channel.", result.error);
    }
  }];
}

RCT_REMAP_METHOD(getChannel, sidOrUniqueName:(NSString *)sidOrUniqueName sid_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  TwilioChatClient *client = [[RCTTwilioChatClient sharedManager] client];
    [[client channelsList] channelWithSidOrUniqueName:sidOrUniqueName completion:^(TCHResult *result, TCHChannel *channel) {
        if (result.isSuccessful) {
            resolve([RCTConvert TCHChannel:channel]);
        }
        else {
            reject(@"not-found", [NSString stringWithFormat:@"Channel could not be found with sid/uniquieName: %@.", sidOrUniqueName], result.error);
        }
    }];
}

#pragma mark Channel Instance Methods

RCT_REMAP_METHOD(setAttributes, sid:(NSString *)sid attributes:(NSDictionary *)attributes attributes_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [RCTTwilioChatChannels loadChannelFromSid:sid :^(TCHResult *result, TCHChannel *channel) {
        if (result.isSuccessful) {
            [channel setAttributes:attributes completion:^(TCHResult *result) {
                if (result.isSuccessful) {
                    resolve(@[@TRUE]);
                }
                else {
                    reject(@"set-attributes-error", @"Error occurred while attempting to setAttributes on channel.", result.error);
                }
            }];
        }
        else {
            reject(@"set-attributes-error", @"Error occurred while attempting to setAttributes on channel.", result.error);
        }
    }];
}

RCT_REMAP_METHOD(setFriendlyName, sid:(NSString *)sid friendlyName:(NSString *)friendlyName attributes_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [RCTTwilioChatChannels loadChannelFromSid:sid :^(TCHResult *result, TCHChannel *channel) {
        if (result.isSuccessful) {
            [channel setFriendlyName:friendlyName completion:^(TCHResult *result) {
                if (result.isSuccessful) {
                    resolve(@[@TRUE]);
                }
                else {
                    reject(@"set-friendlyName-error", @"Error occured while attempting to setFriendlyName on channel.", result.error);
                }
            }];
        }
        else {
            reject(@"set-friendlyName-error", @"Error occured while attempting to setFriendlyName on channel.", result.error);
        }
    }];
}

RCT_REMAP_METHOD(setUniqueName, sid:(NSString *)sid uniqueName:(NSString *)uniqueName attributes_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [RCTTwilioChatChannels loadChannelFromSid:sid :^(TCHResult *result, TCHChannel *channel) {
        if (result.isSuccessful) {
            [channel setUniqueName:uniqueName completion:^(TCHResult *result) {
                if (result.isSuccessful) {
                    resolve(@[@TRUE]);
                }
                else {
                    reject(@"set-friendlyName-error", @"Error occured while attempting to setUniqueName on channel.", result.error);
                }
            }];
        }
        else {
            reject(@"set-friendlyName-error", @"Error occured while attempting to setUniqueName on channel.", result.error);
        }
    }];
}

RCT_REMAP_METHOD(join, sid:(NSString *)sid join_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [RCTTwilioChatChannels loadChannelFromSid:sid :^(TCHResult *result, TCHChannel *channel) {
        if (result.isSuccessful) {
            [channel joinWithCompletion:^(TCHResult *result) {
                if (result.isSuccessful) {
                    resolve(@[@TRUE]);
                }
                else {
                    reject(@"join-channel-error", @"Error occured while attempting to join the channel.", result.error);
                }
            }];
        }
        else {
            reject(@"join-channel-error", @"Error occured while attempting to join the channel.", result.error);
        }

    }];
}

RCT_REMAP_METHOD(declineInvitation, sid:(NSString *)sid decline_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [RCTTwilioChatChannels loadChannelFromSid:sid :^(TCHResult *result, TCHChannel *channel) {
        if (result.isSuccessful) {
            [channel declineInvitationWithCompletion:^(TCHResult *result) {
                if (result.isSuccessful) {
                    resolve(@[@TRUE]);
                }
                else {
                    reject(@"decline-invitation-error", @"Error occured while attempting to decline the invitation to join the channel.", result.error);
                }
            }];
        }
        else {
            reject(@"decline-invitation-error", @"Error occured while attempting to decline the invitation to join the channel.", result.error);
        }
    }];
}

RCT_REMAP_METHOD(leave, sid:(NSString *)sid leave_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [RCTTwilioChatChannels loadChannelFromSid:sid :^(TCHResult *result, TCHChannel *channel) {
        if (result.isSuccessful) {
            [channel leaveWithCompletion:^(TCHResult *result) {
                if (result.isSuccessful) {
                    resolve(@[@TRUE]);
                }
                else {
                    reject(@"leave-channel-error", @"Error occured while attempting to leave the channel.", result.error);
                }
            }];
        }
        else {
            reject(@"leave-channel-error", @"Error occured while attempting to leave the channel.", result.error);
        }
    }];
}

RCT_REMAP_METHOD(destroy, sid:(NSString *)sid destroy_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [RCTTwilioChatChannels loadChannelFromSid:sid :^(TCHResult *result, TCHChannel *channel) {
        if (result.isSuccessful) {
            [channel destroyWithCompletion:^(TCHResult *result) {
                if (result.isSuccessful) {
                    resolve(@[@TRUE]);
                }
                else {
                    reject(@"destroy-channel-error", @"Error occurred while attempting to delete channel.", result.error);
                }
            }];
        }
        else {
            reject(@"destroy-channel-error", @"Error occurred while attempting to delete channel.", result.error);
        }
    }];
}

RCT_REMAP_METHOD(getUnconsumedMessagesCount, sid:(NSString *)sid unconsumedCount_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [RCTTwilioChatChannels loadChannelFromSid:sid :^(TCHResult *result, TCHChannel *channel) {
        if (result.isSuccessful) {
            [channel getUnconsumedMessagesCountWithCompletion:^(TCHResult *result, NSUInteger count) {
                if (result.isSuccessful) {
                    resolve(@(count));
                }
                else {
                    reject(@"get-unconsumed-messages-count-error", @"Error occured while attempting to get unconsumed messages count.", result.error);
                }
             }];
        }
        else {
            reject(@"get-unconsumed-messages-count-error", @"Error occured while attempting to get unconsumed messages count.", result.error);
        }
    }];
}

RCT_REMAP_METHOD(getMessagesCount, sid:(NSString *)sid messagesCount_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [RCTTwilioChatChannels loadChannelFromSid:sid :^(TCHResult *result, TCHChannel *channel) {
        if (result.isSuccessful) {
            [channel getMessagesCountWithCompletion:^(TCHResult *result, NSUInteger count) {
                if (result.isSuccessful) {
                    resolve(@(count));
                }
                else {
                    reject(@"get-messages-count-error", @"Error occured while attempting to get channel messages count.", result.error);
                }
             }];
        }
        else {
            reject(@"get-messages-count-error", @"Error occured while attempting to get channel messages count.", result.error);
        }
    }];
}

RCT_REMAP_METHOD(getMembersCount, sid:(NSString *)sid membersCount_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [RCTTwilioChatChannels loadChannelFromSid:sid :^(TCHResult *result, TCHChannel *channel) {
        if (result.isSuccessful) {
            [channel getMembersCountWithCompletion:^(TCHResult *result, NSUInteger count) {
                if (result.isSuccessful) {
                    resolve(@(count));
                }
                else {
                    reject(@"get-members-count-error", @"Error occurred while attempting to get channel members count.", result.error);
                }
            }];

        }
        else {
            reject(@"get-members-count-error", @"Error occurred while attempting to get channel members count.", result.error);
        }
    }];
}



RCT_REMAP_METHOD(typing, sid:(NSString *)sid) {
    [RCTTwilioChatChannels loadChannelFromSid:sid :^(TCHResult *result, TCHChannel *channel) {
        if (result.isSuccessful) {
            [channel typing];
        }
    }];
}

RCT_REMAP_METHOD(getMember, sid:(NSString *)sid identity:(NSString *)identity member_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [RCTTwilioChatChannels loadChannelFromSid:sid :^(TCHResult *result, TCHChannel *channel) {
        if (result.isSuccessful) {
            resolve([RCTConvert TCHMember:[channel memberWithIdentity:identity]]);
        }
        else {
            reject(@"get-member-error", @"Error occured while attempting to get channel member.", result.error);
        }
    }];
}

RCT_REMAP_METHOD(getChannel, sid:(NSString *)sid channel_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  [RCTTwilioChatChannels loadChannelFromSid:sid :^(TCHResult *result, TCHChannel *channel) {
    if (result.isSuccessful) {
      resolve([RCTConvert TCHChannel:channel]);
    }
    else {
      reject(@"get-full-channel-object-error", @"Error occured while attempting to get full channel object.", result.error);
    }
  }];
}



@end

File: android/src/main/java/com/bradbumbalough/RCTTwilioChat/RCTTwilioChatChannels.java
package com.bradbumbalough.RCTTwilioChat;

import android.support.annotation.Nullable;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReadableMap;

import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.twilio.chat.Channel;
import com.twilio.chat.ChannelListener;
import com.twilio.chat.ChannelDescriptor;
import com.twilio.chat.ErrorInfo;
import com.twilio.chat.Channels;
import com.twilio.chat.Member;
import com.twilio.chat.StatusListener;
import com.twilio.chat.Message;
import com.twilio.chat.CallbackListener;
import com.twilio.chat.Paginator;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.lang.Integer;

import org.json.JSONObject;


public class RCTTwilioChatChannels extends ReactContextBaseJavaModule {

    @Override
    public String getName() {
        return "TwilioChatChannels";
    }

    private ReactApplicationContext reactContext;
    private Map<String,ChannelListener> channelListeners = new HashMap<String, ChannelListener>();

    public RCTTwilioChatChannels(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    private void sendEvent(String eventName, @Nullable WritableMap params) {
        this.reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }

    private ChannelListener generateListener(final Channel channel) {
        final String channelSid = channel.getSid();

        ChannelListener listener = new ChannelListener() {
            @Override
            public void onMessageAdded(Message message) {
                WritableMap map = Arguments.createMap();
                map.putString("channelSid", channelSid);
                map.putMap("message", RCTConvert.Message(message));
                sendEvent("chatClient:channel:messageAdded", map);
            }

            @Override
            public void onMessageUpdated(Message message) {
                WritableMap map = Arguments.createMap();
                map.putString("channelSid", channelSid);
                map.putMap("message", RCTConvert.Message(message));

                sendEvent("chatClient:channel:messageChanged", map);
            }

            @Override
            public void onMessageDeleted(Message message) {
                WritableMap map = Arguments.createMap();
                map.putString("channelSid", channelSid);
                map.putMap("message", RCTConvert.Message(message));

                sendEvent("chatClient:channel:messageDeleted", map);
            }

            @Override
            public void onMemberAdded(Member member) {
                WritableMap map = Arguments.createMap();
                map.putString("channelSid", channelSid);
                map.putMap("member", RCTConvert.Member(member));

                sendEvent("chatClient:channel:memberJoined", map);
            }

            @Override
            public void onMemberUpdated(Member member) {
                WritableMap map = Arguments.createMap();
                map.putString("channelSid", channelSid);
                map.putMap("member", RCTConvert.Member(member));

                sendEvent("chatClient:channel:memberChanged", map);
            }

            @Override
            public void onMemberDeleted(Member member) {
                WritableMap map = Arguments.createMap();
                map.putString("channelSid", channelSid);
                map.putMap("member", RCTConvert.Member(member));

                sendEvent("chatClient:channel:memberLeft", map);
            }

            @Override
            public void onTypingStarted(Member member) {
                WritableMap map = Arguments.createMap();
                map.putString("channelSid", channelSid);
                map.putMap("member", RCTConvert.Member(member));

                sendEvent("chatClient:typingStartedOnChannel", map);
            }

            @Override
            public void onTypingEnded(Member member) {
                WritableMap map = Arguments.createMap();
                map.putString("channelSid", channelSid);
                map.putMap("member", RCTConvert.Member(member));

                sendEvent("chatClient:typingEndedOnChannel", map);
            }

            @Override
            public void onSynchronizationChanged(Channel channel) {
                WritableMap map = Arguments.createMap();
                map.putString("channelSid", channelSid);
                map.putString("status", channel.getSynchronizationStatus().toString());

                sendEvent("chatClient:channel:synchronizationStatusChanged", map);
            }
        };

        return listener;
    }

    private Channels channels() {
        return RCTTwilioChatClient.getInstance().client.getChannels();
    }

    private void createListener(Channel channel) {
        String sid = channel.getSid();
        if (!channelListeners.containsKey(sid)) {
            ChannelListener listener = generateListener(channel);
            channel.addListener(listener);
            channelListeners.put(sid, listener);
        }
    }

    public void loadChannelFromSid(final String sid, final CallbackListener<Channel> callback) {
        channels().getChannel(sid, new CallbackListener<Channel>() {
            @Override
            public void onSuccess(final Channel channel) {
                createListener(channel);
                callback.onSuccess(channel);
            };

            @Override
            public void onError(final ErrorInfo errorInfo){
                callback.onError(errorInfo);
            }

        });
    }

    @ReactMethod
    public void getUserChannels(final Promise promise) {
        channels().getUserChannelsList(new CallbackListener<Paginator<ChannelDescriptor>>() {
            @Override
            public void onError(final ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("get-user-channels-error", "Error occurred while attempting to getUserChannels.");
            }

            @Override
            public void onSuccess(final Paginator<ChannelDescriptor> channelPaginator) {
                String uuid = RCTTwilioChatPaginator.setPaginator(channelPaginator);
                promise.resolve(RCTConvert.Paginator(channelPaginator, uuid, "ChannelDescriptor"));
            }
        });
    }

    @ReactMethod
    public void getSubscribedChannels(final Promise promise) {
        List<Channel> list = channels().getSubscribedChannels();
        promise.resolve(RCTConvert.Channels(new ArrayList<Channel>(list)));
    }

    @ReactMethod
    public void getPublicChannels(final Promise promise) {
        channels().getPublicChannelsList(new CallbackListener<Paginator<ChannelDescriptor>>() {
            @Override
            public void onError(final ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("get-public-channels-error", "Error occurred while attempting to getPublicChannels.");
            }

            @Override
            public void onSuccess(final Paginator<ChannelDescriptor> channelDescriptorPaginator) {
                String uuid = RCTTwilioChatPaginator.setPaginator(channelDescriptorPaginator);
                promise.resolve(RCTConvert.Paginator(channelDescriptorPaginator, uuid, "ChannelDescriptor"));
            }
        });
    }

    @ReactMethod
    public void createChannel(ReadableMap options, final Promise promise) {
        final JSONObject attributes = RCTConvert.readableMapToJson(options.getMap("attributes"));
        final String uniqueName = options.getString("uniqueName");
        String friendlyName = options.getString("friendlyName");
        Channel.ChannelType type = (options.getString("type").compareTo("PRIVATE") == 0) ? Channel.ChannelType.PRIVATE : Channel.ChannelType.PUBLIC;

        channels().channelBuilder()
                .withUniqueName(uniqueName)
                .withFriendlyName(friendlyName)
                .withType(type)
                .withAttributes(attributes)
                .build(new CallbackListener<Channel>() {
                    @Override
                    public void onError(final ErrorInfo errorInfo) {
                        super.onError(errorInfo);
                        promise.reject("create-channel-error", "Error occurred while attempting to createChannel.");
                    }

                    @Override
                    public void onSuccess(final Channel newChannel) {
                        promise.resolve(RCTConvert.Channel(newChannel));
                    }
                });
    }

    @ReactMethod
    public void getChannel(String sid, final Promise promise) {
        loadChannelFromSid(sid, new CallbackListener<Channel>() {
            @Override
            public void onError(final ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("get-full-channel-object-error", "Error occured while attempting to get full channel object.");
            }

            @Override
            public void onSuccess(final Channel channel) {
                promise.resolve(RCTConvert.Channel(channel));
            }
        });
    }

    // Channel Instance Methods

    @ReactMethod
    public void setAttributes(String sid, final ReadableMap attributes, final Promise promise) {
        loadChannelFromSid(sid, new CallbackListener<Channel>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("set-attributes-error", "Error occurred while attempting to setAttributes on channel.");
            }

            @Override
            public void onSuccess(Channel channel) {
                JSONObject json = RCTConvert.readableMapToJson(attributes);
                channel.setAttributes(json, new StatusListener() {
                    @Override
                    public void onError(ErrorInfo errorInfo) {
                        super.onError(errorInfo);
                        promise.reject("set-attributes-error", "Error occurred while attempting to setAttributes on channel.");
                    }

                    @Override
                    public void onSuccess() {
                        promise.resolve(true);
                    }
                });
            }
        });
    }

    @ReactMethod
    public void setFriendlyName(String sid, final String friendlyName, final Promise promise) {
        loadChannelFromSid(sid, new CallbackListener<Channel>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("set-friendly-name-error", "Error occurred while attempting to setFriendlyName on channel.");
            }

            @Override
            public void onSuccess(Channel channel) {
                channel.setFriendlyName(friendlyName, new StatusListener() {
                    @Override
                    public void onError(ErrorInfo errorInfo) {
                        super.onError(errorInfo);
                        promise.reject("set-friendly-name-error", "Error occurred while attempting to setFriendlyName on channel.");
                    }

                    @Override
                    public void onSuccess() {
                        promise.resolve(true);
                    }
                });
            }
        });
    }

    @ReactMethod
    public void setUniqueName(String sid, final String uniqueName, final Promise promise) {
        loadChannelFromSid(sid, new CallbackListener<Channel>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("set-unique-name-error", "Error occurred while attempting to setUniqueName on channel.");
            }

            @Override
            public void onSuccess(Channel channel) {
                channel.setUniqueName(uniqueName, new StatusListener() {
                    @Override
                    public void onError(ErrorInfo errorInfo) {
                        super.onError(errorInfo);
                        promise.reject("set-unique-name-error", "Error occurred while attempting to setUniqueName on channel.");
                    }

                    @Override
                    public void onSuccess() {
                        promise.resolve(true);
                    }
                });
            }
        });
    }

    @ReactMethod
    public void join(String sid, final Promise promise) {
        loadChannelFromSid(sid, new CallbackListener<Channel>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("join-error", "Error occurred while attempting to join channel.");
            }

            @Override
            public void onSuccess(Channel channel) {
                channel.join(new StatusListener() {
                    @Override
                    public void onError(ErrorInfo errorInfo) {
                        super.onError(errorInfo);
                        promise.reject("join-error", "Error occurred while attempting to join channel.");
                    }

                    @Override
                    public void onSuccess() {
                        promise.resolve(true);
                    }
                });
            }
        });
    }

    @ReactMethod
    public void declineInvitation(String sid, final Promise promise) {
        loadChannelFromSid(sid, new CallbackListener<Channel>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("decline-invitation-error", "Error occured while attempting to decline the invitation to join the channel.");
            }

            @Override
            public void onSuccess(Channel channel) {
                channel.declineInvitation(new StatusListener() {
                    @Override
                    public void onError(ErrorInfo errorInfo) {
                        super.onError(errorInfo);
                        promise.reject("decline-invitation-error", "Error occured while attempting to decline the invitation to join the channel.");
                    }

                    @Override
                    public void onSuccess() {
                        promise.resolve(true);
                    }
                });
            }
        });
    }

    @ReactMethod
    public void leave(String sid, final Promise promise) {
        loadChannelFromSid(sid, new CallbackListener<Channel>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("leave-channel-error", "Error occurred while attempting to leave channel.");
            }

            @Override
            public void onSuccess(Channel channel) {
                channel.leave(new StatusListener() {
                    @Override
                    public void onError(ErrorInfo errorInfo) {
                        super.onError(errorInfo);
                        promise.reject("leave-channel-error", "Error occurred while attempting to leave channel.");
                    }

                    @Override
                    public void onSuccess() {
                        promise.resolve(true);
                    }
                });
            }
        });
    }

    @ReactMethod
    public void destroy(String sid, final Promise promise) {
        loadChannelFromSid(sid, new CallbackListener<Channel>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("destroy-channel-error", "Error occurred while attempting to delete channel.");
            }

            @Override
            public void onSuccess(Channel channel) {
                channel.destroy(new StatusListener() {
                    @Override
                    public void onError(ErrorInfo errorInfo) {
                        super.onError(errorInfo);
                        promise.reject("destroy-channel-error", "Error occurred while attempting to delete channel.");
                    }

                    @Override
                    public void onSuccess() {
                        promise.resolve(true);
                    }
                });
            }
        });
    }

    @ReactMethod
    public void typing(String sid) {
        loadChannelFromSid(sid, new CallbackListener<Channel>() {
            @Override
            public void onSuccess(Channel channel) {
                channel.typing();
            }
        });
    }

    @ReactMethod
    public void getUnconsumedMessagesCount(String sid, final Promise promise) {
        loadChannelFromSid(sid, new CallbackListener<Channel>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("get-unconsumed-messages-count-error", "Error occurred while attempting to getUnconsumedMessagesCount.");
            }

            @Override
            public void onSuccess(Channel channel) {
                channel.getUnconsumedMessagesCount(new CallbackListener<Long>() {
                    @Override
                    public void onError(ErrorInfo errorInfo) {
                        super.onError(errorInfo);
                        promise.reject("get-unconsumed-messages-count-error", "Error occurred while attempting to getUnconsumedMessagesCount.");
                    }

                    @Override
                    public void onSuccess(Long count) {
                        promise.resolve(Long.toString(count));
                    }
                });
            }
        });
    }

    @ReactMethod
    public void getMessagesCount(String sid, final Promise promise) {
        loadChannelFromSid(sid, new CallbackListener<Channel>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("get-messages-count-error", "Error occured while attempting to get channel messages count.");
            }

            @Override
            public void onSuccess(Channel channel) {
                channel.getMessagesCount(new CallbackListener<Long>() {
                    @Override
                    public void onError(ErrorInfo errorInfo) {
                        super.onError(errorInfo);
                        promise.reject("get-messages-count-error", "Error occured while attempting to get channel messages count.");
                    }

                    @Override
                    public void onSuccess(Long count) {
                        promise.resolve(Long.toString(count));
                    }
                });
            }
        });
    }

    @ReactMethod
    public void getMembersCount(String sid, final Promise promise) {
        loadChannelFromSid(sid, new CallbackListener<Channel>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
                promise.reject("get-members-count-error", "Error occurred while attempting to get channel members count.");
            }

            @Override
            public void onSuccess(Channel channel) {
                channel.getMembersCount(new CallbackListener<Long>() {
                    @Override
                    public void onError(ErrorInfo errorInfo) {
                        super.onError(errorInfo);
                        promise.reject("get-members-count-error", "Error occurred while attempting to get channel members count.");
                    }

                    @Override
                    public void onSuccess(Long count) {
                        promise.resolve(Long.toString(count));
                    }
                });
            }
        });
    }

    @ReactMethod
    public void shutdown() {
      channelListeners.clear();
    }


    @ReactMethod
    public void getMember(String channelSid, final String identity, final Promise promise) {
        loadChannelFromSid(channelSid, new CallbackListener<Channel>() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                promise.reject("get-member-error", "Error occured while attempting to get channel member");
            }

            @Override
            public void onSuccess(Channel channel) {
                try {
                    Member m = channel.getMembers().getMember(identity);
                    promise.resolve(RCTConvert.Member(m));
                }
                catch (Exception e) {
                    e.printStackTrace();
                    promise.reject("get-member-error", "Error occured while attempting to get channel member.");
                }
            }
        });
    }
}
